pkg/ebpf/c/bpf_telemetry.h

```diff
@@ -0,0 +1,112 @@
+        int helper_indx = -1;                                                                             \
+        int errno_slot;                                                                                   \
+        long is4dot14; \
+        LOAD_CONSTANT("is4dot14", is4dot14); \
+        errno_ret = fn(dst, sz, src);                                                                     \
+        if ((is4dot14) && (errno_ret < 0)) {                                                                              \
+            unsigned long telemetry_program_id;                                                           \
+            LOAD_CONSTANT("telemetry_program_id_key", telemetry_program_id);                              \
+            helper_err_telemetry_t *entry =                                                               \
+                bpf_map_lookup_elem(&helper_err_telemetry_map, &telemetry_program_id);                    \
+            if (entry) {                                                                                  \
+                if (IS_PROBE_READ(fn)) {                                                                  \
+                    helper_indx = read_indx;                                                              \
+                } else if (IS_PROBE_READ_USER(fn)) {                                                      \
+                    helper_indx = read_user_indx;                                                         \
+                } else if (IS_PROBE_READ_KERNEL(fn)) {                                                    \
+                    helper_indx = read_kernel_indx;                                                       \
+                }                                                                                         \
+                errno_slot = errno_ret * -1;                                                              \
+                if (errno_slot >= T_MAX_ERRNO) {                                                          \
+                    errno_slot = T_MAX_ERRNO - 1;                                                         \
+                }                                                                                         \
+                errno_slot &= (T_MAX_ERRNO - 1);                                                          \
+                if (helper_indx >= 0) {                                                                   \
+                    __sync_fetch_and_add(&entry->err_count[(helper_indx * T_MAX_ERRNO) + errno_slot], 1); \
+                }                                                                                         \
+            }                                                                                             \
+        }                                                                                                 \
+    } while (0)
+
+#define MAP_UPDATE(map, key, val, flags) \
```

Comment:
Would it be possible to not use all uppercase names at the least here? It would even better if we didn't have to rename the functions. Can we instead redefine the functions in `bpf_helpers.h`? e.g.:
```c
static int (*_bpf_map_update_elem)(void* map, void* key, void* value, unsigned long long flags) = (void*)BPF_FUNC_map_update_elem;
```
and then here:
```c
#define bpf_map_update_elem(map, key, val, flags) \
  map_update_with_telemetry(_bpf_map_update_elem, map, key, val, flags)
```

Reply:
This was the initial approach. @brycekahle suggested it might be better to rename the functions because
1) The macro invocation is slightly different, in that the actual function takes a reference to a map, whereas the macro does not. This may cause confusion.
2) In certain cases we might want to directly call the helper. For example, in many places bpf_probe_read is called in a loop. In those cases, using the wrapper macro may be too expensive in terms of instruction count.

Reply:
1. That can be remedied by not using a macro and using an inlined function instead
2. That is a good point. I would argue we should rename the functions entirely then to indicate they are different from the regular helpers, e.g. `bpf_probe_read_with_telemetry` or something similar, so we are clear what the difference is. Just `MAP_UPDATE` doesn't really tell me what the difference between this and the regular helper is.

Reply:
1. inlining functions will not work with this approach because we are using the "rewriting" effect of macros to construct the telemetry key for the map in question. https://github.com/DataDog/datadog-agent/blob/860c98c42ababd8c70652f491faa4493895a794a/pkg/ebpf/c/bpf_telemetry.h#L45
I do not see how we can achieve the same effect with an inlined function.

2. I agree with this. I will make the changes.